### PR TITLE
Implement shopConstraint in getCombinationIds()

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Implement shopConstraint in getCombinationIdsAction. Also introduce dedicated CQRS query for same action which reuses the ProductCombinationQueryBuilder instead of relying straight on repository
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | 
| Related PRs       | 
| How to test?      | TBD
| Possible impacts? |

### CODE REVIEW

File: src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php [hunk 1]
```diff
@@ -397,23 +408,32 @@ public function deleteByProductId(ProductId $productId, ShopConstraint $shopCons
      *
      * @return CombinationId[]
      */
-    public function getCombinationIds(ProductId $productId, ShopConstraint $shopConstraint): array
-    {
+    public function getCombinationIds(
+        ProductId $productId,
+        ShopConstraint $shopConstraint,
+        ?ProductCombinationFilters $filters = null
```
Comment (zuk3975) [inline src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php]: not entirely sure about this, it smells abit.
We already require shop_id in filters and then we also have shopConstraint and they are kinda both used to determine the shop. (so one of them feels redundant).
We also use the shop_id in the searchQueryBuilder, so it would seem like we could just have the filters (and require shop_id inside them) without using shopConstraint in this case :thinking: (query builder also uses shop_id from filters and it has strict interface, so we cannot really add additional ShopConstraint param there and I don't think it would be worth it).
Maybe we could build the ShopConstraint inside the filters? Or have it nullable, so that when shop_id = null we would use Constraint::AllShops? (we will have a case for specific prices, where we will need to get combinations for all shops too.).
wdyt @jolelievre

Comment (zuk3975) [reply src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php]: Oh I think I should be able to use ShopCriteriaInterface once its merged from this PR https://github.com/PrestaShop/PrestaShop/pull/30495 so the shopConstraint indeed won't be needed anymore in the method argument, but it will be there in filters.

Comment (jolelievre) [reply src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php]: @zuk3975 the PR just got merged so you should be able to rebase yours Also now that the shop constraint can be built by using `ShopFilters` as a base I think `ProductCombinationFilters` should extend `ShopFilters` and I also think it shouldn't handle the shopId in its generated filterId anymore
Actually it was probably overkill in the first place since the `admin_filter` table already has a shop column,

File: admin-dev/themes/new-theme/js/pages/product/services/paginated-combinations-service.ts [hunk 2]
```diff
@@ -111,6 +111,6 @@ export default class PaginatedCombinationsService implements PaginationServiceTy
   }
 
   private getFilterId(): string {
-    return `product_combinations_${this.productId}_${this.shopId}`;
```
Comment (FabienPapet) [inline admin-dev/themes/new-theme/js/pages/product/services/paginated-combinations-service.ts:114]: shop is no longer needed ? 🤔

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/services/paginated-combinations-service.ts:114]: No, adding it in the first place wasn't needed, the idea was to rely on a filter ID depending on shop ID so that it is associated with shop context. But the filter's DB storage already handles an id_shop column so it was overkill

File: src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php [hunk 3]
```diff
@@ -426,7 +436,7 @@ public function getCombinationIds(ProductId $productId, ShopConstraint $shopCons
             ->andWhere('pas.id_product = :productId')
             ->andWhere($qb->expr()->in('pas.id_shop', ':shopIds'))
             ->setParameter('shopIds', $shopIds, Connection::PARAM_INT_ARRAY)
-            ->setParameter('productId', $productId->getValue())
+            ->setParameter('productId', $productId)
```
Comment (jolelievre) [inline src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php]:
```suggestion
            ->setParameter('productId', $productId->getValue())
```
I think we need the integer value here

File: src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php [hunk 4]
```diff
@@ -88,24 +89,33 @@ class CombinationMultiShopRepository extends AbstractMultiShopObjectModelReposit
...
+     * @param ProductCombinationQueryBuilder $combinationQueryBuilder
      */
     public function __construct(
         Connection $connection,
         string $dbPrefix,
         CombinationValidator $combinationValidator,
         AttributeRepository $attributeRepository,
-        ProductMultiShopRepository $productRepository
+        ProductMultiShopRepository $productRepository,
+        ProductCombinationQueryBuilder $combinationQueryBuilder
```
Comment (jolelievre) [inline src/Adapter/Product/Combination/Repository/CombinationMultiShopRepository.php]: I think you can revert these changes this since it's not used anymore

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            ->fetchAllAssociative()
+        ;
+
+        $combinationIds = [];
+
+        foreach ($results as $result) {
+            $combinationIds[] = new CombinationId((int) $result['id_product_attribute']);
+        }
+
+        return $combinationIds;
```
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php]: Just a cosmetic suggestion, but you should be able to replace this ill an array_map

File: src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php [hunk 6]
```diff
@@ -65,12 +66,9 @@ public function buildFilters(Filters $filters = null)
         }
 
         $productId = $this->getProductId();
-        $shopId = $this->getShopId();
-        $filterId = ProductCombinationFilters::generateFilterId($productId, $shopId);
         $filterParameters['filters']['product_id'] = $productId;
-        $filterParameters['filters']['shop_id'] = $shopId;
 
-        return new ProductCombinationFilters($filterParameters, $filterId);
+        return new ProductCombinationFilters($this->buildShopConstraint(), $filterParameters);
```
Comment (jolelievre) [inline src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php:71]: Actually, since you extend `AbstractFiltersBuilder` you should be able to sue the protected field `$this->shopConstraint` directly, which is based on the provided shopConstraint in the builder's process

It is correctly filled here:
https://github.com/PrestaShop/PrestaShop/blob/efa87169070319d565c5a665d2add7e01778abfc/src/PrestaShopBundle/Controller/ArgumentResolver/FiltersBuilderResolver.php#L67-L76

for all potential builders, and it's correctly fetched from the new request attributes storing the shop constrait. It's better to rely on this one since it's supposed to replace the ShopContext in the long term (we'll be able to control the shop context via GET parameters via this some time in the future)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php [hunk 7]
```diff
@@ -348,23 +347,17 @@ public function getListAction(int $productId, ProductCombinationFilters $combina
...
     {
-        /** @var CombinationRepository $repository */
-        $repository = $this->get('prestashop.adapter.product.combination.repository.combination_repository');
-
-        $combinationIds = $repository->getCombinationIds(new ProductId($productId), $filters);
-        $data = [];
-        foreach ($combinationIds as $combinationId) {
-            $data[] = $combinationId->getValue();
-        }
+        $combinationIds = $this->getQueryBus()->handle(new GetCombinationIds($filters));
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php]: You will need to transform the `$filters` a bit here since your query requires:
- **productId**
- **shopConstraint**
- the other options like filters are optional but should be extracted from the $filters object

File: src/PrestaShopBundle/Resources/config/services/adapter/product/_combination.yml [hunk 8]
```diff
@@ -72,7 +72,7 @@ services:
       - '@doctrine.dbal.default_connection'
       - '%database_prefix%'
       - '@prestashop.adapter.product.combination.validate.combination_validator'
-      - '@prestashop.core.grid.query_builder.product_combination'
+      - '@PrestaShop\PrestaShop\Core\Grid\Query\ProductCombinationQueryBuilder'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/product/_combination.yml]: Not needed anymore I think

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/adapter/product/_combination.yml]:
```suggestion
```

File: src/PrestaShopBundle/Resources/config/services/adapter/product/_combination.yml [hunk 9]
```diff
@@ -85,6 +85,7 @@ services:
       - '@prestashop.adapter.product.combination.validate.combination_validator'
       - '@prestashop.adapter.attribute.repository.attribute_repository'
       - '@prestashop.adapter.product.repository.product_multi_shop_repository'
+      - '@PrestaShop\PrestaShop\Core\Grid\Query\ProductCombinationQueryBuilder'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/product/_combination.yml]:
```suggestion
```

File: tests/Unit/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilderTest.php [hunk 10]
```diff
@@ -59,7 +59,7 @@ public function testBuildFilters()
         $builder->setConfig(['request' => $this->buildRequestMock($productId, $shopId)]);
 
         $builtFilters = $builder->buildFilters();
-        $this->assertEquals(ProductCombinationFilters::generateFilterId($productId, $shopId), $builtFilters->getFilterId());
+        $this->assertEquals(ProductCombinationFilters::generateFilterId($productId), $builtFilters->getFilterId());
```
Comment (jolelievre) [inline tests/Unit/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilderTest.php:62]: It makes me realize we may have missed some unit tests for `ProductCombinationFilters::generateFilterId` even if the implementation evolved no tests to adapt so this part as probably forgotten we should add some simple unit tests

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+      | product1SBlue  |
+      | product1MWhite |
+      | product1MBlack |
+    When I search product "product1" combination ids by following search criteria:
+      | criteria | value |
+      | page     | 2     |
+      | limit    | 5     |
+    Then I should see following paginated combination ids of product "product1":
+      | id reference  | combination name       | reference | attributes          | impact on price | quantity | is default |
+      | product1MBlue | Size - M, Color - Blue |           | [Size:M,Color:Blue] | 0               | 0        | false      |
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]:
```suggestion
      | id reference  |
      | product1MBlue |
```

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]: This step is only supposed to check for IDs (or references)

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+      | product1MBlue  |
+
+  Scenario: I can filter combinations by default combination
+    Given product "product1" combination ids search criteria is set to defaults
+    When I search product "product1" combination ids by following search criteria:
+      | criteria   | value |
+      | is default | true  |
+    Then I should see following filtered combination ids of product "product1":
+      | id reference   | combination name        | reference | attributes           | impact on price | quantity | is default |
+      | product1SWhite | Size - S, Color - White |           | [Size:S,Color:White] | 0               | 0        | true       |
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]:
```suggestion
      | id reference   |
      | product1SWhite |
```

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+    When I search product "product1" combination ids by following search criteria:
+      | criteria   | value |
+      | is default | false |
+    Then I should see following filtered combination ids of product "product1":
+      | id reference   | combination name        | reference | attributes           | impact on price | quantity | is default |
+      | product1SBlack | Size - S, Color - Black |           | [Size:S,Color:Black] | 0               | 0        | false      |
+      | product1SBlue  | Size - S, Color - Blue  |           | [Size:S,Color:Blue]  | 0               | 0        | false      |
+      | product1MWhite | Size - M, Color - White |           | [Size:M,Color:White] | 0               | 0        | false      |
+      | product1MBlack | Size - M, Color - Black |           | [Size:M,Color:Black] | 0               | 0        | false      |
+      | product1MBlue  | Size - M, Color - Blue  |           | [Size:M,Color:Blue]  | 0               | 0        | false      |
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]:
```suggestion
      | id reference   |
      | product1SBlack |
      | product1SBlue  |
      | product1MWhite |
      | product1MBlack |
      | product1MBlue  |
```

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+      | impact on price | 10   |
+    And I update combination "product1SBlack" stock with following details:
+      | delta quantity  | 50 |
+    And I update combination "product1SBlack" with following values:
+      | reference       | CCC |
+    When I search product "product1" combinations list by following search criteria:
+      | criteria  | value           |
+      | order by  | impact on price |
+      | order way | asc             |
+    Then I should see following combinations in paginated list of product "product1":
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]: Is it necessary to check the list of full data as well in this scenario? Or is it to ensure that both queries return the same thing?

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

I know it's still a WIP I just had a few suggestions
Also, do you plan on handling multishop use cases and behat tests in this PR as well? I think it could

File: src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php (hunk 6 above)
Comment (zuk3975) [reply src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php:71]: but it will make the shopId coming from request useless now as it still takes shop constraint from context

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature (hunk 14 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]: its just for convenience, so we can see its sorted correctly and yeah that both return the same (if I leave ids only, we won't see the price, therefore visually it won't be as clear.), but idk, I could get rid of them if you think its not worth it

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php (hunk 5 above)
Comment (zuk3975) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php]: at first i done this with array_map + array_column, but it wasn't looking better. But yeah, only array_map and manual casting should be better than loop 😄

File: src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php (hunk 6 above)
Comment (jolelievre) [reply src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php:71]: Oh you're right, forget it then! We'll refacto this later

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php [hunk 15]
```diff
@@ -348,23 +347,25 @@ public function getListAction(int $productId, ProductCombinationFilters $combina
     /**
      * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))")
      *
-     * @param int $productId
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php:351]: This param is still present it should be kept, or all must be removed

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+                'limit' => $query->getLimit(),
+                'offset' => $query->getOffset(),
+                'orderBy' => $query->getOrderBy(),
+                'sortOrder' => $query->getOrderWay(),
+                'filters' => $filters,
+            ]
+        );
+
+        $results = $this->productCombinationQueryBuilder
+            ->getSearchQueryBuilder($searchCriteria)
```
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php:79]:
```suggestion
            ->getSearchQueryBuilder($searchCriteria)
            ->select('pas.id_product_attribute')
```

You can override the select part so that the query only returns the required data

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature (hunk 14 above)
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Product/Combination/get_combination_ids.feature]: No it was just to be sure, so it makes sense to use both steps (one that checks the full data and the other only the IDs) however you need to clean the table in the steps dedicated to IDs since they are not used

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php (hunk 16 above)
Comment (zuk3975) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php:79]: yeah been there, but all selects are still needed as otherwise filters fails

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php (hunk 15 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php:351]: idk why git shows it this way, its there, maybe outdated change.

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php (hunk 16 above)
Comment (zuk3975) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php:79]: actually not all parts are needed, but only price, it fails to order by it and throws "nonUniqueFieldNameException" when ordering by price. 
So I think its enough if I select id_product attribute and price here

Comment (zuk3975) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php:79]: this is what I ended up with, I think its fair
[8cfbb28](https://github.com/PrestaShop/PrestaShop/pull/30585/commits/8cfbb28eb5be626674ae9b78fb3666dc60b64b36)

Comment (jolelievre) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php:79]: Yep, seems fine!

File: src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+            ->select('pas.id_product_attribute')
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        return array_map(static function (array $result): CombinationId {
+            return new CombinationId($result['id_product_attribute']);
```
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php]:
```suggestion
            return new CombinationId((int) $result['id_product_attribute']);
```

Comment (jolelievre) [reply src/Core/Domain/Product/Combination/QueryHandler/GetCombinationIdsHandler.php]: Seems like this is what makes the CI go red

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

**Review by lartist — APPROVED** [review]

**Comment by jolelievre** [comment]
No need for QA this is a technical and partial PR, it will be tested as a whole once the feature is finished
